### ISSUE TITLE
WARNING: ThreadSanitizer: Swift access race

### ISSUE THREAD

**Issue by robertmryan** [onevcat/Kingfisher#1085]
### Issue Description

If I turn on ThreadSanitizer with Xcode 10.1 and Swift 4.2 with an app that is using `UIImageView` extension, I’m getting numerous “Swift access race” errors if an new request happens before the prior one for the same image view has finished

> WARNING: ThreadSanitizer: Swift access race

#### What

TSAN is reporting errors such as:

<img width="841" alt="screen shot 2019-01-07 at 3 15 10 pm" src="https://user-images.githubusercontent.com/1530768/50799217-fd408c80-128f-11e9-9bbd-7e43b6dafe28.png">

Or

<img width="842" alt="screen shot 2019-01-07 at 3 13 57 pm" src="https://user-images.githubusercontent.com/1530768/50799239-06c9f480-1290-11e9-93fa-5c593d82107f.png">

Or

<img width="842" alt="screen shot 2019-01-07 at 3 13 27 pm" src="https://user-images.githubusercontent.com/1530768/50799255-0e899900-1290-11e9-8e9b-a2fe85b4861a.png">

#### Reproduce

```swift
import UIKit
import Kingfisher

class ViewController: UIViewController {
    @IBOutlet weak var tableView: UITableView!    
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1000
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CustomCell", for: indexPath) as! CustomCell
        let url = ...
        cell.customImageView.kf.setImage(with: url, options: [
            .processor(DownsamplingImageProcessor(size: cell.customImageView.bounds.size)),
            .scaleFactor(UIScreen.main.scale),
            .cacheOriginalImage
            ])
        return cell
    }
}
```

The problem appears to manifest itself if I scroll quickly enough that prior images do not have a chance to finish before the cell (and its image view) are reused.

Even simpler, this MCVE without `DownsamplingImageProcessor` seems to reproduce the problem with some consistency, too:

```swift
//  ViewController.swift

import UIKit
import Kingfisher

class ViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let urls = [
            "https://images-assets.nasa.gov/image/KSC-20171024-PH_ULA01_0001/KSC-20171024-PH_ULA01_0001~orig.jpg",
            "https://images-assets.nasa.gov/image/201407020003HQ/201407020003HQ~orig.tif",
            "https://images-assets.nasa.gov/image/sts063-s-007/sts063-s-007~orig.jpg"
        ].map { URL(string: $0)! }
        
        let group = DispatchGroup()

        for url in urls {
            group.enter()
            self.imageView.kf.setImage(with: url) { result in
                group.leave()
            }
        }
        
        group.notify(queue: .main) {
            print("done")
        }
    }    
}
```

Obviously, the second code snippet, above, isn’t a practical example (why bother updating the image view with three different images), but just is an even simpler example of how to manifest the race condition.

**Comment by robertmryan** [comment]
By the way, the second example doesn’t reproduce the problem as consistently as my first example with 1000 images, but I wager if I add more URLs to the array, it started producing it with greater frequency, e.g.,

```swift
let urls = [
    "https://images-assets.nasa.gov/image/KSC-20171024-PH_ULA01_0001/KSC-20171024-PH_ULA01_0001~orig.jpg",
    "https://images-assets.nasa.gov/image/201407020003HQ/201407020003HQ~orig.tif",
    "https://images-assets.nasa.gov/image/sts063-s-007/sts063-s-007~orig.jpg",
    "https://images-assets.nasa.gov/image/PIA23001/PIA23001~orig.jpg",
    "https://images-assets.nasa.gov/image/PIA22977/PIA22977~orig.jpg",
    "https://images-assets.nasa.gov/image/KSC-20180803-PH_BOE01_0005/KSC-20180803-PH_BOE01_0005~orig.jpg",
    "https://images-assets.nasa.gov/image/PIA22990/PIA22990~orig.jpg",
    "https://images-assets.nasa.gov/image/KSC-20180803-PH_BOE01_0001/KSC-20180803-PH_BOE01_0001~orig.jpg",
    "https://images-assets.nasa.gov/image/AFRC2018-0287-072/AFRC2018-0287-072~orig.jpg"
].map { URL(string: $0)! }
```

And, by the way, this second example actual has the “Swift access race” in a slightly different spot, namely:

<img width="838" alt="screen shot 2019-01-07 at 3 51 09 pm" src="https://user-images.githubusercontent.com/1530768/50800251-53afca00-1294-11e9-87c1-af6b1f0ae249.png">

But the consistency in both of these examples is that the image view has its image set asynchronously before the prior image had a chance to finish.

**Comment by onevcat** [comment]
Thanks for reporting this. I'd investigate it soon.

**Comment by onevcat** [comment]
@robertmryan 
I guess it should be fixed in #1089 
Can you try to confirm it in your enviroment? Thank you!
